### PR TITLE
Add error when emoji are invalid

### PR DESCRIPTION
**Pull request by cehali**
# Summary
Add error when emoji are invalid

### CODE REVIEW

**Comment by cehali** [comment]
image
image
image
image

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 1]
```diff
@@ -26,13 +26,13 @@ export const EmojiForm = ({hideHeader, className, notifications, onDenyClick, se
     [notifications, enteredCode],
   );
 
-  const isInputValid = enteredCode.length === SECURITY_CODE_LENGTH && soleAddress && isValidCode(enteredCode, soleAddress);
+  const isInputValid = enteredCode.length === SECURITY_CODE_LENGTH && soleAddress !== undefined && isValidCode(enteredCode, soleAddress);
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Move
```ts
enteredCode.length === SECURITY_CODE_LENGTH && soleAddress !== undefined
```
to `isValidCode`

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 2]
```diff
@@ -26,13 +26,13 @@ export const EmojiForm = ({hideHeader, className, notifications, onDenyClick, se
...
 
   useEffect(() => {
     if (isInputValid) {
       hideHeader && hideHeader();
       ensureNotNull(soleAddress, Error, 'No matching keys');
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Move to isValidCode

File: universal-login-react/src/ui/Notifications/EmojiPlaceholders.tsx [hunk 3]
```diff
@@ -32,12 +33,17 @@ export const EmojiPlaceholders = ({enteredCode, onEmojiClick, className}: EmojiP
     return [...emojis, ...placeholders];
   };
 
+  const isEmojiInputCompleteButInvalid = enteredCode.length === SECURITY_CODE_LENGTH && !isEmojiInputValid;
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiPlaceholders.tsx]: Move to the same file as `isValidCode` and add unit tests.

Comment (cehali) [reply universal-login-react/src/ui/Notifications/EmojiPlaceholders.tsx]: Done

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx (hunk 2 above)
Comment (cehali) [reply universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Done

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx (hunk 1 above)
Comment (cehali) [reply universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Done

**Comment by cehali** [comment]
image
image
image
image
image
image
image
image

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 4]
```diff
@@ -26,13 +25,13 @@ export const EmojiForm = ({hideHeader, className, notifications, onDenyClick, se
     [notifications, enteredCode],
   );
 
-  const isInputValid = enteredCode.length === SECURITY_CODE_LENGTH && soleAddress && isValidCode(enteredCode, soleAddress);
+  const isInputValid = isValidCode(enteredCode, soleAddress);
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Can be inlined now

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 5]
```diff
@@ -26,13 +25,13 @@ export const EmojiForm = ({hideHeader, className, notifications, onDenyClick, se
...
 
   useEffect(() => {
     if (isInputValid) {
       hideHeader && hideHeader();
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: hideHeader?.()

File: universal-login-commons/lib/core/utils/securityCodes.ts [hunk 6]
```diff
@@ -31,11 +31,17 @@ export const generateCode = (publicKey: string): number[] => {
     .map((e) => e & MASK);
 };
 
-export const isValidCode = (code: number[], publicKey: string) => {
-  const expectedCode = generateCode(publicKey);
-  return deepEqual(expectedCode, code);
+export const isValidCode = (code: number[], publicKey?: string) => {
+  if (publicKey) {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/securityCodes.ts]: No related test added

File: universal-login-commons/test/core/utils/securityCodes.ts [hunk 7]
```diff
@@ -105,6 +105,25 @@ describe('UNIT: security codes', () => {
...
 
+  describe('isCodeSufficientButInvalid', () => {
+    it('correct code and valid length', () => {
+      const encodedAddress = generateCode(mockedAddress);
+      expect(isCodeSufficientButInvalid(encodedAddress, mockedAddress)).to.be.false;
+    });
+
+    it('wrong code and valid length', () => {
+      const codeKeyboard = generateCodeWithFakes(mockedAddress);
+      const wrongCodePickedByUser = codeKeyboard.splice(0, 6);
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/securityCodes.ts]: non-deterministric

**Pull request merged**